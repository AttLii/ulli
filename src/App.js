import { Ulli } from "./components/Ulli";

export default function App() {
  return (
    <>
      <Ulli
        ListElement="ol"
        className="hi"
        itemClassName="hiho"
        list={["item1", "item2"]}
        itemFunc={(item, i) => (
          <p>
            {i} {item}
          </p>
        )}
      />
    </>
  );
}