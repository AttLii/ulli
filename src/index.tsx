import React, {ElementType, ReactNode} from "react"

type Props<T> = {
  list: T[];
  itemFunc: (item: T, index: number) => ReactNode
  className?: string
  itemClassName?: string
  ListElement?: ElementType
  ItemElement?: ElementType
};

export const Ulli = <T,>({
  list,
  itemFunc,
  className = undefined,
  itemClassName = undefined,
  ItemElement = "li",
  ListElement = "ul"
}: Props<T>) => (
  list.length > 0 ? (
    <ListElement className={className}>
      {list.map((item, i) => (
        <ItemElement key={i} className={itemClassName}>
          {itemFunc(item, i)}
        </ItemElement>
      ))}
    </ListElement>
  ) : null
)


/* => (
  list.length > 0 ? (
    <ListElement className={className}>
      {list.map((item, i) => (
        <ItemElement key={i} className={itemClassName}>
          {itemFunc(item, i)}
        </ItemElement>
      ))}
    </ListElement>
  ) : null
) */