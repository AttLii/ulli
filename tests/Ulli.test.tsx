import { render } from '@testing-library/react';
import React from "react"

import { Ulli } from '../src/index';

test("doesn't crash", () => {
  render(<Ulli />);
});

test("doesn't render anything without list-prop", () => {
  const { container } = render(<Ulli />)
  expect(container.firstChild).toMatchSnapshot();
});

test("doesn't render anything with empty list-prop", () => {
  const { container } = render(<Ulli list={[]} />)
  expect(container.firstChild).toMatchSnapshot();
});

test("renders only ul > li without itemFn-prop", () => {
  const { container } = render(<Ulli list={["item1", "item2"]} />)
  expect(container.firstChild).toMatchSnapshot();
});

test("renders itemFunc's result as each <li> elements children", () => {
  const itemFunc = () => "hi"

  const { container } = render(
    <Ulli
      list={["item1", "item2"]}
      itemFunc={itemFunc}
    />
  )
  expect(container.firstChild).toMatchSnapshot();
});

test("itemFunc has access to each item and current index", () => {
  const { container } = render(
    <Ulli
      list={["item1", "item2"]}
      itemFunc={(item, i) => <div>{item} {i}</div>}
    />
  )
  expect(container.firstChild).toMatchSnapshot();
});

test("component accepts className and itemClassName prop, that is applied to elements", () => {
  const { container } = render(
    <Ulli
      className="app__list"
      itemClassName="app__list-item"
      list={["item1", "item2"]}
      itemFunc={(item, i) => <div>{item} {i}</div>}
    />
  )
  expect(container.firstChild).toMatchSnapshot();
});

test("component accepts ListElement and ItemElement prop, that can be a react component", () => {
  const { container } = render(
    <Ulli
      list={["item1", "item2"]}
      ListElement={({ children, ...rest }) => <div {...rest}>{children}</div>}
      ItemElement={({ children, ...rest }) => <button {...rest}>{children}</button>}
    />
  )
  expect(container.firstChild).toMatchSnapshot();
});

test("component accepts ListElement and ItemElement prop, that can be a string of html element", () => {
  const { container } = render(
    <Ulli
      list={["item1", "item2"]}
      ListElement="div"
      ItemElement="button"
    />
  )
  expect(container.firstChild).toMatchSnapshot();
});