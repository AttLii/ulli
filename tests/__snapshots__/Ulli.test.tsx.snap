// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component accepts ListElement and ItemElement prop, that can be a react component 1`] = `
<div>
  <button />
  <button />
</div>
`;

exports[`component accepts ListElement and ItemElement prop, that can be a string of html element 1`] = `
<div>
  <button />
  <button />
</div>
`;

exports[`component accepts className and itemClassName prop, that is applied to elements 1`] = `
<ul
  class="app__list"
>
  <li
    class="app__list-item"
  >
    <div>
      item1
       
      0
    </div>
  </li>
  <li
    class="app__list-item"
  >
    <div>
      item2
       
      1
    </div>
  </li>
</ul>
`;

exports[`doesn't render anything with empty list-prop 1`] = `null`;

exports[`doesn't render anything without list-prop 1`] = `null`;

exports[`itemFunc has access to each item and current index 1`] = `
<ul>
  <li>
    <div>
      item1
       
      0
    </div>
  </li>
  <li>
    <div>
      item2
       
      1
    </div>
  </li>
</ul>
`;

exports[`renders itemFunc's result as each <li> elements children 1`] = `
<ul>
  <li>
    hi
  </li>
  <li>
    hi
  </li>
</ul>
`;

exports[`renders only ul > li without itemFn-prop 1`] = `
<ul>
  <li />
  <li />
</ul>
`;
